### ISSUE TITLE
Idea: Testing `fail!`ures with `TestResult::must_fail`

### ISSUE THREAD

**Issue by japaric** [BurntSushi/quickcheck#18]
Currently, quickcheck doesn't provide a way to test properties that must trigger a task failure to pass. For example, if you want to test that bound checking is working, you would write the following property:

``` rust
#[quickcheck]
fn out_of_bounds(length: uint, index: uint) -> TestResult {
    let v = Vec::from_elem(length, 0u8);

    if index < length {
        TestResult::discard()
    } else {
        // v[index] ?
    }
}
```

But you are missing a piece: You need some function that returns `TestResult::passed()` if `v[index]` triggers a task failure, or `TestResult::failed()` if `v[index]` succeeded.

That's why I propose adding the `must_fail` function to the library, where `must_fail` is defined as:

``` rust
use quickcheck::TestResult;
use std::comm;
use std::io::ChanWriter;
use std::task::TaskBuilder;

// This should be a `TestResult` static method
fn must_fail<T: Send>(f: proc(): Send -> T) -> TestResult {
    // A "/dev/null" sink for the stderr of the child task
    let (tx, _) = comm::channel();
    let dev_null = box ChanWriter::new(tx) as Box<Writer + Send>;

    let failed =
        TaskBuilder::new().
            // Redirect the stderr of the child task to "/dev/null"
            stderr(dev_null).
            try(f).
            is_err();

    TestResult::from_bool(failed)
}
```

This function provides the missing functionality to complete the `out_of_bounds` property:

``` rust
#[quickcheck]
fn out_of_bounds(length: uint, index: uint) -> TestResult {
    let v = Vec::from_elem(length, 0u8);

    if index < length {
        TestResult::discard()
    } else {
        must_fail(proc() v[index])
    }
}
```

Which now can be tested:

```
running 1 test
test out_of_bounds ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
```

And if you wrongly implement the test, for example by reversing the `index < length` condition, then the test would fail as expected:

```
running 1 test
test out_of_bounds ... FAILED

failures:

---- out_of_bounds stdout ----
        task 'out_of_bounds' failed at '[quickcheck] TEST FAILED. Arguments: (1, 0)', src/lib.rs:83



failures:
    out_of_bounds

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured
```

@BurntSushi What do you think? Should this functionality be added to quickcheck?
